### PR TITLE
Add search debounce feature

### PR DESCRIPTION
**Pull request by fnkhz**
Required for implementing pbek/QOwnNotes#2335

### CODE REVIEW

File: qmarkdowntextedit.cpp [hunk 1]
```diff
@@ -111,6 +111,12 @@ void QMarkdownTextEdit::setLineNumbersOtherLineColor(QColor color) {
     _lineNumArea->setOtherLineColor(std::move(color));
 }
 
+void QMarkdownTextEdit::setDebounceDelay(uint debounceDelay)
```
Comment (pbek) [inline qmarkdowntextedit.cpp:114]: Should maybe be more descriptive, like `setSearchWidgetDebounceDelay`

File: qplaintexteditsearchwidget.h [hunk 2]
```diff
@@ -44,7 +46,10 @@ class QPlainTextEditSearchWidget : public QWidget {
     int _currentSearchResult;
     QList<QTextEdit::ExtraSelection> _searchExtraSelections;
     QColor selectionColor;
+    QTimer _debounceTimer;
+    QString searchTerm_;
```
Comment (pbek) [inline qplaintexteditsearchwidget.h:50]: Is there a reason for the trailing `_` instead of a leading one?

File: qplaintexteditsearchwidget.cpp [hunk 3]
```diff
@@ -395,6 +431,11 @@ void QPlainTextEditSearchWidget::setSearchMode(SearchMode searchMode) {
     ui->modeComboBox->setCurrentIndex(searchMode);
 }
 
+void QPlainTextEditSearchWidget::setDebounceDelay(uint debounceDelay)
+{
+    _debounceTimer.setInterval(debounceDelay);
```
Comment (pbek) [inline qplaintexteditsearchwidget.cpp:436]: You are casting a uint to int here...

**Comment by pbek** [comment]
But it seems to work nice! Great job!

**Comment by Waqar144** [comment]
Not sure what this PR is doing. Are you just delaying the search a little bit? I don't see the point in doing that

**Comment by pbek** [comment]
> I don't see the point in doing that

you have time to enter more characters and the ui isn't blocking in the meantime (if search is "slow")...

**Comment by Waqar144** [comment]
> > I don't see the point in doing that
> 
> you have time to enter more characters and the ui isn't blocking in the meantime (if search is "slow")...

That's just moving the "slowness" further away. The core issue is bad search implementation which should be fixed instead so that its not so slow.

The search should start getting slow if someone is searching in a note worth a few megabytes at least. In those cases we can stop the search if it crosses a certain threshold, say 200ms or something. So, something like:

```
while (doSearch) {
   if (currentRunTimeInMS >= 300) {
      break;
   }
}
```

This has another benefit that it allows you to cancel the search with some modifications. But another UI might be needed to notify the user that search is taking too long and we are stopping it till you press enter.

**Comment by Waqar144** [comment]
Also, the name "debounce" is really opaque. I don't really see how debounce correlates to what this feature is doing.

File: qplaintexteditsearchwidget.h (hunk 2 above)
Comment (fnkhz) [reply qplaintexteditsearchwidget.h:50]: No, will change it.

File: qplaintexteditsearchwidget.cpp (hunk 3 above)
Comment (fnkhz) [reply qplaintexteditsearchwidget.cpp]: Updated; using explicit `static_cast`.

**Comment by fnkhz** [comment]
@Waqar144 The debounce term seems to be pretty common in  the JS world. What name should we use instead ?

**Comment by fnkhz** [comment]
> But it seems to work nice! Great job!
Thanks ! Glad it works fine.

**Comment by fnkhz** [comment]
@Waqar144 Also noticed the issues with the search; could also work on that if you haven't started yet.

**Comment by Waqar144** [comment]
> @Waqar144 Also noticed the issues with the search; could also work on that if you haven't already started.

Sure.

**Comment by Waqar144** [comment]
I think #147 is the better way to go. The users for whom the search is really slow can just press enter when they are done typing what they want to search.

**Comment by pbek** [comment]
But isn't this PR much more convenient... I'd rather wait 200ms (which will be a setting in QON, default 0) then having to press enter every time (even if pressing enter is faster if you only do it once, but it's counter-intuitive imho) the whole web is used to wait a few 100ms for a search input.. I also would want that in the script repository search... even I forget to press enter there (I don't want to search every time because that spams the github api)

**Comment by Waqar144** [comment]
Sure. No problem.

But I'd rather go for something that automatically triggers instead of having the user set an option. One can check the delay and turn on debouncing if needed.

**Comment by pbek** [comment]
> But I'd rather go for something that automatically triggers instead of having the user set an option. One can check the delay and turn on debouncing if needed.

Wasn't that what we were aiming for? wait 0sec by default and let the user increase that value... Or did you propose something different? 😁

**Comment by Waqar144** [comment]
> wait 0sec by default and let the user increase that value

No. My point => don't let the user change this. Decide ourselves based on what is happening.

**Comment by pbek** [comment]
> Decide ourselves based on what is happening.

Decide on a value (e.g. 300ms) or decide what makes the search on slow computers with big files slow?

**Comment by Waqar144** [comment]
> > Decide ourselves based on what is happening.
> 
> Decide on a value (e.g. 300ms) or decide what makes the search on slow computers with big files slow?

Yes, that prevents a user from accidentally setting a bad value in the options.

But it's not as simple to implement, a lot of things need to be restructured to allow doing this properly.

File: qplaintexteditsearchwidget.h (hunk 2 above)
Comment (pbek) [reply qplaintexteditsearchwidget.h:50]: Now both are trailing instead of leading...

**Comment by pbek** [comment]
For now I'll set a search delay of 300ms for text with more than 200k characters in QOwnNotes (and 0 below). We can remove that later on...

**Pull request merged**